Added Customizing Look and Functionality section (#149)
smartdevicelink/sdl_app_library_guides: docs/Customizing Look and Functionality/Customizing Help Prompts/index.md

# Customizing Help Prompts
You are able to customize the head unit's help system with your own information.

## Help Menu

> If the head unit supports displaying a custom help menu, you can customize this menu with your own title and menu options. This menu is commonly used to assist users with knowing what voice commands are available. If you don't customize these options, then the head unit's default title and/or items will be used.

Reviewer: ```suggestion
You can customize the help menu with your own title and/or menu options. If you don't customize these options, then the head unit's default menu will be used.
```

// Set up the menu items
SDLVRHelpItem *item1 = [[SDLVRHelpItem alloc] initWithText:<#Help item name such as "Show Artists"#> image: <#A previously uploaded image or nil#> position: 1];
SDLVRHelpItem *item2 = [[SDLVRHelpItem alloc] initWithText:<#Help item name such as "Shuffle All"#> image: <#A previously uploaded image or nil#> position: 2];
setGlobals.vrHelp = @[item1, item2];

[self.sdlManager sendRequest:setGlobals withResponseHandler:^(SDLRPCRequest *request, SDLRPCResponse *response, NSError *error) {
    if (error != nil) {
        // Something went wrong
    }

    // The help prompt is updated
}];
```

##### Swift
```swift
let setGlobals = SDLSetGlobalProperties()
setGlobals.vrHelpTitle = <#Custom help title string such as: "What Can I Say?"#>

// Set up the menu items
let item1 = SDLVRHelpItem(text:<#Help item name such as "Show Artists"#>, image: <#A previously uploaded image or nil#>, position: 1)
let item2 = SDLVRHelpItem(text:<#Help item name such as "Shuffle All"#>, image: <#A previously uploaded image or nil#>, position: 2)
setGlobals.vrHelp = [item1, item2];

sdlManager.send(request: setGlobals) { (request, response, error) in

>     if error != nil {

Author: Changed to `if let`, using `guard` doesn't work well here because while you can do:

```swift
guard error != nil else { return }
```

`error` is not unwrapped from optional here, so you'd actually have to do:

```swift
guard error != nil {
  if let error = error {
  }

  return
}
```

Instead you can just do `if let`

    // The help prompt is updated
}
```
!@

@![android, javaSE, javaEE]
```java
// TODO
```
!@

## Help Prompt
In SDL, the user is able to request assistance with using your app by saying "Help" on head units that support voice recognition. This will display the help menu discussed above, but it can also speak a custom spoken text-to-speech response to the user.

@![iOS]
##### Objective-C
```objc
SDLSetGlobalProperties *setGlobals = [[SDLSetGlobalProperties alloc] init];
setGlobals.helpPrompt = SDLTTSChunk *response = [SDLTTSChunk textChunksFromString:<#Your custom help prompt#>];

[self.sdlManager sendRequest:setGlobals withResponseHandler:^(SDLRPCRequest *request, SDLRPCResponse *response, NSError *error) {
    if (error != nil) {
        // Something went wrong
    }

    // The help prompt is updated
}];
```

##### Swift
```swift
let setGlobals = SDLSetGlobalProperties()
setGlobals.helpPrompt = SDLTTSChunk.textChunks(from: <#Your custom help prompt#>)

sdlManager.send(request: setGlobals) { (request, response, error) in
    if error != nil {
        // Something went wrong
    }

    // The help prompt is updated
}
```
!@

@![android, javaSE, javaEE]
```java
// TODO
```
!@

> ## Timeout Prompt

Reviewer: Should probably add another section "Clearing the Help Menu" with instructions on how to remove the custom help menu.

> ## Timeout Prompt

Reviewer: I feel like "Timeout Prompt" is out of place in this section. Maybe it should be moved to it's own section?

# Customizing Help Prompts

> On some head units it is possible to display a customized help menu or speak a custom command if the user asks for help while using your app. The help menu is commonly used to let users know what voice commands are available however it can also be customized to help your user navigate the app or let them know what features are available. 

Reviewer: ```suggestion
On some head units it is possible to display a customized help menu or speak a custom command if the user asks for help while using your app. The help menu is commonly used to let users know what voice commands are available, however, it can also be customized to help your user navigate the app or let them know what features are available. 
```

## Configuring the Help Menu
You can customize the help menu with your own title and/or menu options. If you don't customize these options, then the head unit's default menu will be used.

> If you wish to use an image, you should check the @![iOS]`sdlManager.systemCapabilityManager.defaultMainWindowCapability.imageFields`!@@![android, javaSE, javaEE]`// TODO`!@ for an `imageField.name` of `vrHelpItem` to see if that image is supported. If `vrHelpItem` is in the `imageFields` array, then it can be used. You will need to then upload the image using the file manager before using it in the RPC below. See the [Uploading Images guide](Other SDL Features/Uploading Images) for more information.

Reviewer: ```suggestion
If you wish to use an image, you should check the @![iOS]`sdlManager.systemCapabilityManager.defaultMainWindowCapability.imageFields`!@@![android, javaSE, javaEE]`// TODO`!@ for an `imageField.name` of `vrHelpItem` to see if that image is supported. If `vrHelpItem` is in the `imageFields` array, then it can be used. You will then need to upload the image using the file manager before using it in the request. See the [Uploading Images](Other SDL Features/Uploading Images) section for more information.
```

@![iOS]
##### Objective-C
```objc
SDLSetGlobalProperties *setGlobals = [[SDLSetGlobalProperties alloc] init];

> setGlobals.vrHelpTitle = <#Custom help title string such as: "What Can I Say?"#>

Reviewer: ```suggestion
setGlobals.vrHelpTitle = <#Custom help title string such as: "What Can I Say?"#>;
```

// Set up the menu items
SDLVRHelpItem *item1 = [[SDLVRHelpItem alloc] initWithText:<#Help item name such as "Show Artists"#> image: <#A previously uploaded image or nil#> position: 1];
SDLVRHelpItem *item2 = [[SDLVRHelpItem alloc] initWithText:<#Help item name such as "Shuffle All"#> image: <#A previously uploaded image or nil#> position: 2];
setGlobals.vrHelp = @[item1, item2];

[self.sdlManager sendRequest:setGlobals withResponseHandler:^(SDLRPCRequest *request, SDLRPCResponse *response, NSError *error) {
    if (error != nil) {

>         // Something went wrong

Reviewer: Should this `return` to match the Swift examples?

Author: done

    }

    // The help menu is updated
}];
```

##### Swift
```swift
let setGlobals = SDLSetGlobalProperties()
setGlobals.vrHelpTitle = <#Custom help title string such as: "What Can I Say?"#>

// Set up the menu items
let item1 = SDLVRHelpItem(text:<#Help item name such as "Show Artists"#>, image: <#A previously uploaded image or nil#>, position: 1)
let item2 = SDLVRHelpItem(text:<#Help item name such as "Shuffle All"#>, image: <#A previously uploaded image or nil#>, position: 2)
setGlobals.vrHelp = [item1, item2];

sdlManager.send(request: setGlobals) { (request, response, error) in

>     if let error = error {

Reviewer: ```suggestion
    guard error == nil else {
```

Author: This doesn't let you do anything with the error (it'll still be optional) if something went wrong

        // Something went wrong
        return;
    }

    // The help menu is updated
}
```
!@

@![android, javaSE, javaEE]
```java
// TODO
```
!@

## Configuring the  Help Prompt
On head units that support voice recognition, a user can request assistance by saying "Help." In addition to displaying the help menu discussed above a custom spoken text-to-speech response can be spoken to the user.

@![iOS]
##### Objective-C
```objc
SDLSetGlobalProperties *setGlobals = [[SDLSetGlobalProperties alloc] init];

> setGlobals.helpPrompt = SDLTTSChunk *response = [SDLTTSChunk textChunksFromString:<#Your custom help prompt#>];

Reviewer: ```suggestion
setGlobals.helpPrompt = [SDLTTSChunk textChunksFromString:<#Your custom help prompt#>];
```

[self.sdlManager sendRequest:setGlobals withResponseHandler:^(SDLRPCRequest *request, SDLRPCResponse *response, NSError *error) {
    if (error != nil) {
        // Something went wrong
    }

    // The help prompt is updated
}];
```

##### Swift
```swift
let setGlobals = SDLSetGlobalProperties()
setGlobals.helpPrompt = SDLTTSChunk.textChunks(from: <#Your custom help prompt#>)

sdlManager.send(request: setGlobals) { (request, response, error) in

>     if let error = error {

Reviewer: ```suggestion
    guard error == nil else {
```

Author: See above comment

        // Something went wrong
        return
    }

    // The help prompt is updated
}
```
!@

@![android, javaSE, javaEE]
```java
// TODO
```
!@

## Configuring the Timeout Prompt
If you display any sort of popup menu or modal interaction that has a timeout – such as an alert, interaction, or slider – you can create a custom text-to-speech response that will be spoken to the user in the event that a timeout occurs.

@![iOS]
##### Objective-C
```objc
SDLSetGlobalProperties *setGlobals = [[SDLSetGlobalProperties alloc] init];

> setGlobals.timeoutPrompt = SDLTTSChunk *response = [SDLTTSChunk textChunksFromString:<#Your custom help prompt#>];

Reviewer: ```suggestion
setGlobals.timeoutPrompt = [SDLTTSChunk textChunksFromString:<#Your custom help prompt#>];
```

[self.sdlManager sendRequest:setGlobals withResponseHandler:^(SDLRPCRequest *request, SDLRPCResponse *response, NSError *error) {
    if (error != nil) {
        // Something went wrong
    }

    // The timeout prompt is updated
}];
```

##### Swift
```swift
let setGlobals = SDLSetGlobalProperties()
setGlobals.timeoutPrompt = SDLTTSChunk.textChunks(from: <#Your custom help prompt#>)

sdlManager.send(request: setGlobals) { (request, response, error) in

>     if let error = error {

Reviewer: ```suggestion
    guard error == nil else {
```

Author: See above comment

// Reset the menu icon and title
SDLResetGlobalProperties *resetGlobals = [[SDLResetGlobalProperties alloc] initWithProperties:@[SDLGlobalPropertyMenuIcon, SDLGlobalPropertyMenuName]];

// Reset the spoken prompts
SDLResetGlobalProperties *resetGlobals = [[SDLResetGlobalProperties alloc] initWithProperties:@[SDLGlobalPropertyHelpPrompt, SDLGlobalPropertyTimeoutPrompt]];

[self.sdlManager sendRequest:resetGlobals withResponseHandler:^(SDLRPCRequest *request, SDLRPCResponse *response, NSError *error) {
    if (error != nil) {
        // Something went wrong
    }

    // The global properties are reset
}];
```

##### Swift
```swift
// Reset the help menu
let resetGlobals = SDLResetGlobalProperties(properties: [.voiceRecognitionHelpItems, .voiceRecognitionHelpTitle])

// Reset the menu icon and title
let resetGlobals = SDLResetGlobalProperties(properties: [.menuIcon, .menuName])

// Reset the spoken prompts
let resetGlobals = SDLResetGlobalProperties(properties: [.helpPrompt, .timeoutPrompt])
sdlManager.send(request: resetGlobals) { (req, res, err) in

>     if let error = error {

Reviewer: ```suggestion
    guard error == nil else {
```

Author: See above comment

// Set up the menu items
let item1 = SDLVRHelpItem(text:<#Help item name such as "Show Artists"#>, image: <#A previously uploaded image or nil#>, position: 1)
let item2 = SDLVRHelpItem(text:<#Help item name such as "Shuffle All"#>, image: <#A previously uploaded image or nil#>, position: 2)
setGlobals.vrHelp = [item1, item2];

sdlManager.send(request: setGlobals) { (request, response, error) in
    if let error = error {
        // Something went wrong
        return;
    }

    // The help menu is updated
}
```
!@

@![android, javaSE, javaEE]
```java
// TODO
```
!@

## Configuring the  Help Prompt
On head units that support voice recognition, a user can request assistance by saying "Help." In addition to displaying the help menu discussed above a custom spoken text-to-speech response can be spoken to the user.

@![iOS]
##### Objective-C
```objc
SDLSetGlobalProperties *setGlobals = [[SDLSetGlobalProperties alloc] init];
setGlobals.helpPrompt = [SDLTTSChunk textChunksFromString:<#Your custom help prompt#>];

[self.sdlManager sendRequest:setGlobals withResponseHandler:^(SDLRPCRequest *request, SDLRPCResponse *response, NSError *error) {
    if (error != nil) {

>         // Something went wrong

Reviewer: ```suggestion
        // Something went wrong
        return;
```
